Add porcelain flux jobs command

chu11 opened the pull request:
This is an initial hacked up `flux-queue.py` that I wrote up (built on top of PR #2580) just to get my python bearings.  Example output:

```
JOBID		STATE	TASKS	NAME	USERID	TIME
21357395968	R	1	sleep  	8556	0:00:01
18756927488	R	1	sleep  	8556	0:00:01
16190013440	R	1	sleep  	8556	0:00:01
```

Options are:

```
  -c N, --count N       Limit output to N jobs
  -p, --pending         List pending jobs
  -r, --running         List running jobs
  -i, --inactive        List inactive jobs
  -s, --suppress-header
                        Suppress printing of header line
  -u UID, --userid UID  Limit output to specific userid (Specify "all" for all
                        users)
```

So options aren't all that different from `flux job list` for the time being, but it:

A) lists task count
B) lists run time
C) caps output to 1000 jobs (random max I thought was acceptable)
D) defaults to showing running / pending jobs.

Notes:

- should priority be output by default?  Will average user care?  I suppose we need an option for this atleast.
- it's a little annoying having to build an rpc for `job-info.list` and passing it into `h.rpc()`.  Perhaps I need to make a binding for `flux_job_list()`.
- is it kosher to call `raw.flux_job_statetostr()` in such a tool?  I'm not sure if calling via `raw` should only be isolated to bindings?

perhaps should have a brainstorming / discussion of what we would like the defaults to be or what initial user options we'd like.  @garlick had some ideas in #2416 

I dislike the `--running`, `--pending`, `--inactive` options, as well as `-u UID`.  I think we should leave them in there, but I'll throw out the following as starters:

- `-a` - conceptually similar to `ps -a`, show all jobs associated with a user.  This would be shorthand for `--pending --running --inactive`.

- `-A` - conceptually similar to `ps -A`, show everything, inactive, all users, etc.  This would be shorthand for `--pending --running --inactive -u all`

grondo commented:
Some quick thoughts, though I think you're right a quick brainstorming wouldn't hurt.

 * This tool isn't limited to, or even really listing, the job "queue". So I wonder if a simpler name is in order? What about `flux jobs`?
 * Instead of different switches for various states, what about a single `--states=NAME,...` to limit output to selected states. This kind of interface is more extensible and a better UI. Later, the interface could be enhanced to support somthing like `--states=all,^run` or something to select all jobs *except* those that are running.
 * For output fields, we had talked about supporting formatted output strings, perhaps using Python f-strings, with a `-o, --format` option. Then there could just be a default format registered within the script, or perhaps by configuration (users could also supply their own format strings under `~/.flux` perhaps.)
 * For user readable RUNTIME output I prefer floating point output with units. This way you get more information by default for short running jobs. E.g. (1.52s, 33.5m, 1.2hr, 2.5d). It is also typically a bit shorter than `DD:HH:MM:SS`. This is just my preference, and it could always be added as an option later.
 * As a porcelain tool, the tool should support usernames, not UIDs.
 * The "default" number of jobs returned could be set by configuration, but a sane default might be something much smaller, like 25, with a message that the output was truncated (if there is a way to determine that)

The even simpler tool alluded to in #2416 could maybe be under a different name, summarizing more than just jobs.

grondo commented:
> it's a little annoying having to build an rpc for job-info.list and passing it into h.rpc(). Perhaps I need to make a binding for flux_job_list().

This is a good idea. See merlin use case from mailing list...

chu11 commented:
> Instead of different switches for various states, what about a single --states=NAME,... to limit output to selected states. This kind of interface is more extensible and a better UI. Later, the interface could be enhanced to support somthing like --states=all,^run or something to select all jobs except those that are running.

Sounds like a great idea.  I was thinking of something similar as well, b/c I got tired of typing `--inactive --pending --running`.

> The "default" number of jobs returned could be set by configuration, but a sane default might be something much smaller, like 25, with a message that the output was truncated (if there is a way to determine that)

Not at the moment, but we can get the length of the pending/running/inactive queues.  Could simply return them with the job listing.

chu11 commented:
> it's a little annoying having to build an rpc for job-info.list and passing it into h.rpc(). Perhaps I need to make a binding for flux_job_list().

Re-pushed with some minor tweaks (from some of comments above) + a `flux_job_list()` python binding.  But how to do this was not as easy as I had hoped.

Try 1
- Make a binding to `flux_job_list()`, which returns a future.  I return `Future(future_handle)` to the user.  No biggie I think.
- User wants to call `future.get()` to get the list of jobs from the response.  Doh!  The `RPC` class from the python bindings supports `get()`, not the `Future` class.
- Move the `get()` and `get_str()` functions from the `RPC` to `Future` class.  Ok?   Not really the right idea I think?
- Considered if we could have `RPC` initialized only with a future, but not currently doable given how constructor is implemented (I think).

Try 2
- An alternate approach would be to simply re-implement the RPC call using the python binding for `flux.job.rpc()`.  Works fine, except it duplicates code ;-(  (i.e. if `job-info.list` service changes, gotta update in two locations)

Try 3
- I thought part of the issue is we don't have a `flux_job_list_get()` C function, which eliminates this problem for most of our bindings.  So I add one.
- But the prototype is `flux_job_list_get (flux_future_t *, json_t **)`, and the `json_t **` makes the binding not as simple as desired.

So I've gone with the second option for now.

chu11 commented:
was messing around with this more today and added a few new things:

- `-a` option for all of a user's jobs (equivalent to `--state=pending,running,inactive`)
- `-A` for all jobs (equivalent to `--state=pending,running,inactive -u all`)
- runtime output defaults to flux standard duration format

So for example, the default output looks like this now:

```
JOBID		STATE	TASKS	NAME	USERID	TIME
26860322816	I	1	sleep  	achu	3.1s
23605542912	I	1	sleep  	achu	2.1s
20384317440	I	1	sleep  	achu	1.1s
```

- a `--format option`, that allows user to specify their own format using python fstring format

For example, the default output is output like this now:

```
        fmt=("{id}\t{state_single}\t{task_count}\t"
             "{job_name:<7.7}\t{username}\t{runtime_fsd}")
        output_format(fmt, jobs)
```

users can use any of the attributes returned by `job-info`, as can be seen in the `output_format()` function.  I add in a few extras like `username` which converts `userid` to name, and `state_single` is the single character state (vs `state` is full name).  There's several runtime variants for time output format.  The above `runtime_fsd` is the flux standard duration variant.

We can obviously add virtually any number of variants we desire to support a wide range of formatting options.

chu11 commented:
Just re-pushed something that is beginning to approach something that can be merged.  One big remaining todo is I need to write the manpage.

In latest push, I tweaked the default output to be:

```
JOBID		STATE    TASKS  NAME     USER     TIME
23991418880	RUN      1      sleep    achu     0.16s
20954742784	RUN      1      sleep    achu     0.34s
17850957824	RUN      1      sleep    achu     0.53s
```

The primary difference to last time is I now output the full name of the State.  @grondo mentioned it yesterday, and as I wrote up tests, the full name seemed to be better.

I added a whole bunch of tests in `t2800-jobs-cmd.t`.

I also said that this PR would fix #2416 and #1693.  While some of the sub-issues listed in #2416 aren't done, I think we've ticked off the primary check box of that issue.  And I believe `--format` scratches off #1693.

chu11 commented:
Finally fixed up all those lgtms and travis test failures.  So I think this is approaching completion, need to add a manpage still.

chu11 commented:
and I've added a manpage, so I've removed the `[WIP]` for now.

@garlick brought up yesterday that he was not the biggest fan of the name of this cmd being `flux-jobs` b/c we already have a `flux-job`. 

@grondo didn't like `flux-queue` b/c it didn't quite map correctly, since we can also list inactive jobs.

Naming wise I couldn't think of anything really good yesterday.  Only thing I could come up with was possibly `flux-list`?  But I don't even like that.

garlick commented:
Er, now I'm regretting opening my mouth without first having a counter proposal.  I'm stumped!  I think `flux jobs` may be the best idea so far, so unless someone else has a last minute great idea, I say full speed ahead.

chu11 commented:
after discussion this afternoon and reading through some comments internally provided, just listing out a few nice things to have for the future.  I'll write up an issue once this PR is merged.

- supporting completion status as a possible output type (probably requires #2590 to be done first, where data for a job transition won't become available until after data for that transition is read)
- a "verbose" default output, possibilities include priority, start time, endtime (for inactive), (once supported) completion status, and maybe other stuff
- a compaction option, that will list identical jobs for the same user on a single line, hypothetically if job-name, task count, username, job-state, and maybe 1 or 2 other things are the same, batch em up into the same line.

garlick commented on doc/man1/flux-jobs.adoc:
```diff
@@ -0,0 +1,100 @@
+
+OPTIONS
+-------
+*-a*::
```
drop extra "is"

garlick commented on doc/man1/flux-jobs.adoc:
```diff
@@ -0,0 +1,100 @@
+// flux-help-include: true
+FLUX-JOBS(1)
+==============
```
I found the  formatted man page seems to be indented far to the right.  It seems to be fixed by making the length of the `====` underline match the length of the title.  (This is with asciidoctor as formatter)

garlick commented on doc/man1/flux-jobs.adoc:
```diff
@@ -0,0 +1,100 @@
+'--state=pending,running,inactive'.
+
+*-A*::
+List all jobs from all users, including completed jobs is in the inactive state.
```
"completed jobs in the inactive state" seems redundant and confusing.  How about just "inactive jobs"?

chu11 commented on doc/man1/flux-jobs.adoc:
```diff
@@ -0,0 +1,100 @@
+*-c, --count*'=N'::
+Limit output to N jobs (default 1000)
+
+*-s, --states*'=STATES'::
```
I see your point. `--lists`? `--categories`?  I'll see if something better comes to mind tonight.

grondo replied:
Since this is a porcelain tool, maybe it doesn't matter that the states presented here don't match the flux internal job states? We should use whatever makes sense for the tool itself. That being said, for consistency it might be nice for the names of states in the `--states` option to match the states presented in the `STATE` column in the tool's output.

chu11 replied:
It just occurred to me, `--states` on the command line refers to the job list the job is on (pending, running, inactive), but `STATE` column refers to the job state (depend, sched, run, cleanup, inactive).  That could be confusing ... hmmm ...

grondo replied:
To me it seems obvious what the `--states` option does. Maybe "pending" and "running" could be considered virtual states for the purposes of this tool, where `"pending" == "sched,depend"` and `"running" == "run,cleanup"` (then rename "inactive" to "complete" or something). I don't think users should have to be aware of the internal implementation of the `job-info` module to use `flux jobs` (or even that there is a `job-info` module for the casual user).

Another idea would be to make the option even more generic, like a `--filter` or `--include` option. This might make it clearer that the names passed into the option don't necessarily directly correspond to "official" job states.

garlick replied:
Idea: maybe keep `--states` and add  a small section towards the end of the man page that describes states and virtual states, which can be referenced from the option description up top, e.g.

`--states=STATES`::
...where STATES is a comma-separated list of job states, or virtual job states such as _pending_, _running_ , or _inactive_ - see JOB STATES below.

Such as
> JOB STATES
> Jobs may be observed to pass through five job states in Flux: DEPEND, SCHED, RUN, CLEANUP, and INACTIVE (see Flux RFC 21).   For convenience and clarity, some options accept the following virtual job states:  "pending", an alias for DEPEND,SCHED; "running", an alias for RUN,CLEANUP.

grondo replied:
That seems like a reasonable solution. We may also want to add "active" as an alias for "pending,running" for completeness.

chu11 replied:
```
 --states=STATES::
...where STATES is a comma-separated list of job states, or virtual job states such as pending, running , or inactive - see JOB STATES below.
```

So would a user be able to specify `--states=DEPEND` or `--states=CLEANUP`?  Just don't have that functionality at the moment.

garlick replied:
My vote is yes.  Just noted that the C api introduces these virtual states:
```c
enum job_list_flags {
    FLUX_JOB_LIST_ALL = 0,
    FLUX_JOB_LIST_PENDING = 1,
    FLUX_JOB_LIST_RUNNING = 2,
    FLUX_JOB_LIST_INACTIVE = 4,
};
```
It would be better IMHO if the API accepted a mask of real states, and the tool took care of translating the virtual states to a state mask.

I'd be OK with this change coming in a later PR since it seems like this is otherwise pretty close.

garlick replied:
We could also "canonicalize" the virutal states in the state enum, by adding
```c
typedef enum {
    FLUX_JOB_NEW                    = 1,
    FLUX_JOB_DEPEND                 = 2,
    FLUX_JOB_SCHED                  = 4,
    FLUX_JOB_RUN                    = 8,
    FLUX_JOB_CLEANUP                = 16,
    FLUX_JOB_INACTIVE               = 32,   // captive end state
    /* Virtual states for convenience
    */
    FLUX_JOB_PENDING = (FLUX_JOB_DEPEND | FLUX_JOB_SCHED),
    FLUX_JOB_RUNNING = (FLUX_JOB_RUN | FLUX_JOB_CLEANUP),
    FLUX_JOB_ACTIVE = (FLUX_JOB_PENDING | FLUX_JOB_RUNNING),
} flux_job_state_t;
```

chu11 replied:
> I'd be OK with this change coming in a later PR since it seems like this is otherwise pretty close.

Agreed, this should be for a future PR.

grondo commented on src/cmd/flux-jobs.py:
```diff
@@ -0,0 +1,202 @@
+            )
+        fmt = (
+            "{id}\t{state:<8.8} {task_count:<6} "
+            "{job_name:<8.8} {username:<8.8} {runtime_fsd}"
```
The default format could use some tweaking IMO. This is very subjective, but at the least:

 * more space for job name (since there is currently extra width available) Maybe 10 characters at least
 * It would be nice if runtime was blank instead of "0s" for "pending" jobs. Actually, maybe TIME could be "time since submission" for pending jobs? (Not sure if that is a good idea or not)
 * I think JOBID is going to need more than "2 tabs" worth of spaces by default. O/w after some amount of time FLUIDs will overflow this space and the header won't line up. Probably won't be an issue in the short term, but I fast forwarded FLUID generator by six days and got the output below. Probably want to use a default width for `{id}` of 18. (Techinically I guess we'd need 20, but by my testing it takes about 1.88 years for the FLUIDs to roll over to 19 digits)
```
$ flux jobs -a
JOBID		STATE    TASKS  NAME     USER     TIME
8697698811117568	INACTIVE 10     hostname grondo   0.08s
8697690472841216	INACTIVE 10     hostname grondo   0.08s
8697681698357248	INACTIVE 10     hostname grondo   0.08s
8697674333159424	INACTIVE 10     hostname grondo   0.08s
8697374104879104	INACTIVE 10     sleep    grondo   10s
```

* It is weird for me to see the username as one of the last fields in the default output format. Again this is totally subjective, but maybe the default format should be `JOBID  USER  NAME STATE NTASKS TIME`.

chu11 replied:
> It is weird for me to see the username as one of the last fields in the default output format.

FWIW, the order I picked was somewhat random :-)  Re-ordered 1 or 2 times b/c a certain order "felt" good.  But we can totally re-order. :-)

grondo commented on t/t2800-jobs-cmd.t:
```diff
@@ -0,0 +1,262 @@
+#!/bin/sh
+
+test_description='Test flux job info service'
```
Update of  `test_description` needed here.

SteVwonder commented on src/cmd/flux-jobs.py:
```diff
@@ -0,0 +1,202 @@
+        if not args.suppress_header:
+            print(
+                "%s\t\t%-8.8s %-6.6s %-8.8s %-8.8s %s"
+                % ("JOBID", "STATE", "TASKS", "NAME", "USER", "TIME")
```
FWIW, this should fix the alignment issues that @grondo pointed out
```python
fmt.format(
id="JOBID",
state="STATE",
task_count="TASKS",
job_name="NAME",
username="USER",
runtime_fsd="TIME"
)
```

chu11 replied:
ahh good idea, i'm clearly not thinking pythonic :-)

SteVwonder commented on src/bindings/python/flux/job.py:
```diff
@@ -98,6 +98,32 @@ def wait(flux_handle, jobid=lib.FLUX_JOBID_ANY):
+# is more of a reimplementation of flux_job_list().  Subtle issues
+# are:
+#
+# - Python Future class does not have get()
```
It does; it's just not explicitly written.  Calling `.get()` on any future will call `flux_future_get` thanks to some magic that @trws wrote in the `Wrapper` class.

chu11 replied:
ahhh, what I meant was `.get_str()` which is only in `rpc.py`

SteVwonder commented on src/bindings/python/flux/job.py:
```diff
@@ -98,6 +98,32 @@ def wait(flux_handle, jobid=lib.FLUX_JOBID_ANY):
+    try:
+        ret = future_handle.get()
+    except EnvironmentError as err:
+        raise err
```
I may be wrong, but I don't think this try/except block buys you anything.  It catches the exception and then re-raises it immediately.  I'd just drop the try/except.

SteVwonder commented on doc/man1/flux-jobs.adoc:
```diff
@@ -0,0 +1,100 @@
+
+The '--format' option can be used to specify an output format to
+flux-jobs(1) using Python f-string formations.  For example, the
+following is the format used for the default format:
```
This is really cool!

SteVwonder commented on doc/man1/flux-jobs.adoc:
```diff
@@ -0,0 +1,100 @@
+*--suppress-header*::
+For default output, do not output column headers.
+
+*-u, --userid*'=UID'::
```
We may want to reserve the `-u` short arg for username.   I suspect users will prefer to do `flux jobs -u herbein1` (or `flux jobs -u $USER`) than `flux jobs -u $(id -u herbein1)`.

chu11 replied:
good idea, perhaps it should just be `--user` and take username or userid?  i.e. assume userid if it's only numbers.

chu11 replied:
actually, `--user` is in `squeue`, which actually takes multiple usernames.  Perhaps could be done with `flux-jobs`, but that's for another day.  I can add that to the issue list.

chu11 replied:
hmmmm, I currently support `--userid=all`, but I probably shouldn't support `--user=all`, b/c a username could be named `all`.  Hmmm.  `--user=0` means all?  not the most obvious.

chu11 replied:
Well, it seems `lsf` just assumes there will never be a user named `all`.

`-u user_name | -u user_group | -u all`

chu11 commented on src/cmd/flux-jobs.py:
```diff
@@ -0,0 +1,202 @@
+        sys.exit(1)
+
+    if args.format:
+        output_format(args.format, jobs)
```
Yeah, b/c it's Python f-strings it's hard to do this.  Since the user can specify virtually any formatting they want.  I think with `squeue` there's a finite set of output options with their `"type[:[.]size]" `format?

trws replied:
I can think of a couple ways we could make this happen, mainly either extracting the variables with a parse `{\w*` or similar, or setting up a mock format that we use to determine which fields were accessed by the formatting process.  Either way this is probably a down-the-road thing but I think it's doable.

chu11 replied:
given stephen's technique below of just generating a header via:

```
fmt.format(
id="JOBID",
state="STATE",
task_count="TASKS",
job_name="NAME",
username="USER",
runtime_fsd="TIME"
)
```

It's perhaps easier than I thought.  Ultimately, if the header doesn't look pretty, we leave it to the user to figure out?

chu11 commented on src/cmd/flux-jobs.py:
```diff
@@ -0,0 +1,202 @@
+        "--format",
+        type=str,
+        metavar="FORMAT",
+        help="Specify output format using Python f-string formation",
```
and I should fix in the manpage too.

trws replied:
Thanks for this, in 3.6 the difference is largely academic, but in 3.8 f-strings get some formatting specifiers that `.format()` can't support, so it's good to be clear what will work.

chu11 commented:
re-pushed with fixes per comments above, new default output is:

```
JOBID              USER     NAME       STATE    NTASKS TIME
44778389504        achu     sleep      INACTIVE 1      3.1s
42429579264        achu     sleep      INACTIVE 1      2.1s
39879442432        achu     sleep      INACTIVE 1      1.1s
```

The two things I didn't fix are:

- the `-u, --userid` option is still there.  I couldn't decide if I should change to `-u, --user` or should there be a `-u, --user` and `--userid`?  I couldn't figure it out this evening.  Maybe I'll figure it out tomorrow.

- I didn't output an empty string when the runtime is `0s`.  Once I saw the output, I just didn't like it :-)

grondo commented:
> the -u, --userid option is still there. I couldn't decide if I should change to -u, --user or should there be a -u, --user and --userid? I couldn't figure it out this evening. Maybe I'll figure it out tomorrow.

A common approach is to allow `--user=[UID|USERNAME]`. First try to resolve the argument as a name, if that fails treat it as a UID. If both fail then generate error message. You can disambiguate UID by prefixing a `+` character [as is done in coreutils](https://www.gnu.org/software/coreutils/manual/coreutils.html#Disambiguating-names-and-IDs).

grondo commented:
> Maybe it should be like a - instead? I implemented it via a new format field runtime_fsd_nozero (of course we can rename it to whatever we want).

Yeah, I would like a `-` as well, but I don't think the `0s` is too bad. Like I said it was just IMO.

grondo commented:
> format is

> `{id:<18} {username:<8.8} {job_name:<10.10} {state:<8.8} {task_count:<6} {runtime_fsd}`

One more suggestion for this round. Integers should be right-justified by default. So I'd change `{task_count:>6}` (just assuming that is how you right justify). I might even do the same for `{id:>18}`, if only because this is how `ps` formats PIDs. But I leave that up to someone else.

chu11 commented:
> One more suggestion for this round. Integers should be right-justified by default. So I'd change {task_count:>6} (just assuming that is how you right justify). I might even do the same for {id:>18}, if only because this is how ps formats PIDs. But I leave that up to someone else.

I think it's fine for task count, but with job id initially:

```
             JOBID USER     NAME       STATE    NTASKS TIME
       22414360576 achu     sleep      RUN           1 0.16s
       19293798400 achu     sleep      RUN           1 0.34s
       16290676736 achu     sleep      RUN           1 0.52s
```

Should look better once there's more jobids in there.

I think `ps` is calculating the max width of an PID integer and adjusting the right justify accordingly.  I'd hate to loop through the job list just for that.  I'm thinking I'll leave the jobid left justified for now and we can debate this later.

garlick commented:
I like the right justified jobid actually :-)

chu11 commented:
> I like the right justified jobid actually :-)

LOL, ok I've been outvoted :-)

chu11 commented:
rebased, squashed, and re-pushed

- Default output is now:

```
             JOBID USER     NAME       STATE    NTASKS TIME
       88130715648 achu     sleep      SCHED         1 -
       91133837312 achu     sleep      SCHED         1 -
       85211480064 achu     sleep      RUN           1 2.9s
       82292244480 achu     sleep      RUN           1 3.1s
```

using

`{id:>18} {username:<8.8} {job_name:<10.10} {state:<8.8} {task_count:>6} {runtime_fsd_hyphen}`

- you'll notice a hyphen for the runtime for a non-running job.  The format specifier for that is `runtime_fsd_hyphen`.  The format specifier `runtime_fsd` would give you `0s` for zero runtime.

- `-u, --user=[USERNAME|UID]` is now the option name for user.  Internally the check is:

```
    if args.user == "all":
        userid = flux.constants.FLUX_USERID_UNKNOWN
    else:
        try:
            userid = pwd.getpwnam(args.user).pw_uid
        except KeyError:
            try:
                userid = int(args.user)
            except ValueError:
                print("invalid user specified", file=sys.stderr)
                sys.exit(1)
```

Not the prettiest, but I think its ok.

Issues I think we agree for later are:

- `--format` has a header
- `--states` includes actual job states
- completion status output
- verbose default output
- compaction option

grondo commented:
Coming out very nicely! Nice work!

garlick commented:
> Coming out very nicely! Nice work!

Yeah!  I gave my official review approval already earlier.  Just wanted to say I'm fine with this going in if @grondo thinks it is ready.

chu11 commented:
hold off on the merge, just spoke to @SteVwonder, I think I can cleanup something in the bindings.

chu11 commented:
I don't think the cleanup is gonna happen, but I did tweak some comments in the bindings.

garlick commented:
Restarted a builder that timed out in ` t1103-apidisconnect`

garlick commented:
Looks good!  Thanks!

I'll set merge-when-passing.
